libethash-cuda/CUDAMiner.cpp

```diff
@@ -478,24 +472,47 @@ void CUDAMiner::search(
             }
 
             // See if we got solutions in this batch
-            uint32_t found_count = std::min((unsigned)buffer->count, SEARCH_RESULTS);
+            uint32_t found_count = buffer->count;
             if (found_count)
             {
-
+                // save the solution(s), We'll submit them after all
+                buffer_cpy = &m_search_buf_cpy[current_index];
+                // the streams have been restarted.
+                if (found_count > SEARCH_RESULTS)
+                    found_count = SEARCH_RESULTS;
+                for (unsigned i = 0; i < found_count; i++)
+                    buffer_cpy->result[i] = ((search_results*)buffer)->result[i];
+                buffer_cpy->count = found_count;
                 buffer->count = 0;
-                uint64_t nonce_base = current_nonce - streams_batch_size;
+            }
+            // restart the stream on the next batch of nonces
+            if (!done)
+                run_ethash_search(
+                    s_gridSize, s_blockSize, stream, buffer, current_nonce, m_parallelHash);
+        }
+        // All streams are running. time for some bookeeping
+        updateHashRate(batch_size, s_numStreams);
+        for (current_index = 0; current_index < s_numStreams; current_index++)
```

Comment:
Why this second loop is ran anyway ? Shouldn't be under condition if(found_count) ?

Reply:
I don't know if you noticed, but there is a fundamental change here from what was there before your change.

Previously, we would copy on return from a stream, relaunch the stream, then report any solutions found by that stream. That would maximize overlap of HOST and GPU processing for that stream, but it would still delay the processing of stream 2, stream 3,...

This does the copies as before, but only processes results after relaunching the last stream without delay. That's what the added loop is about, looking through stream results for solutions.

Reply:
Call me dumb (have the right to) but it's too convoluted for me

* I don't get where `&m_search_buf_cpy[current_index]` is set
* I don't get that `2` as constant
* I don't get why this `buffer_cpy = &m_search_buf_cpy[current_index];` is ran twice

Must admit my judgement is heavily influenced by the difficulty of reading such logic.

>Previously, we would copy on return from a stream, relaunch the stream, then report any solutions found by that stream

That's not the correct order. Previously we waited for a stream, submit solutions if any, and relaunch the stream.

Reply:
> It is set here via the buffer_cpy pointer that points to one of its elements.

You set buffer_cpy there **from** m_search_buf_cpy[current_index] but I can't find where m_search_buf_cpy[current_index] is valued.

Reply:
It's where I showed
```
                buffer_cpy = &m_search_buf_cpy[current_index];
                // the streams have been restarted.
                if (found_count > SEARCH_RESULTS)
                    found_count = SEARCH_RESULTS;
                for (unsigned i = 0; i < found_count; i++)
                    buffer_cpy->result[i] = ((search_results*)buffer)->result[i];
                buffer_cpy->count = found_count;
```

buffer_cpy is a pointer to the search_results data type. It is set to point to one of the elements of m_search_buf_cpy. A couple of lines down we assign values through that pointer, effectively setting an element in m_search_buf_cpy.

Reply:
There are two pointers buffer and buffer_cpy. Should they be renamed to something_ptr? I'll move the definitions to make it clear that they are pointers.

Reply:
Never mind ... I don't get it as a whole anyway.
I give up.

Reply:
Try it and see.